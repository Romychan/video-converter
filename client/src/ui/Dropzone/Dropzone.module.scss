.container {
  width: 100%;
  height: 100%;

  &.active .background {
    display: block;
    z-index: 1;
  }

  &.active .button {
    z-index: 0;
  }
}

.dropzone {
  width: 100%;
  min-height: 415px;
  border: 2px dashed var(--color-border);
  background-color: var(--color-background);
  border-radius: 15px;
  z-index: 5;
  position: relative;
  padding: 60px 30px;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  z-index: 2;

  &.error {
    border: 2px dashed var(--color-red);
  }
}

.background {
  position: fixed;
  top: 0px;
  left: 0px;
  width: 100%;
  height: 100%;
  background: #0000007c;
  z-index: -1;
  display: none;
}

.icon {
  width: 90px;
  height: 65px;
  margin-bottom: 2rem;

  & svg {
    width: 100%;
    height: 100%;
  }
}

.subtitle {
  font-size: 24px;
  margin-bottom: 1rem;
  text-align: center;
}

.text {
  margin-bottom: 1rem;
}

.button {
  font-size: 12px;
  font-family: var(--text-family-regular);
  z-index: 4;
}

.input {
  font-size: 18px;
  width: 100%;
  border: none;
  text-transform: none;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  opacity: 0;
  cursor: pointer;

  &:focus {
    outline: none;
  }
}

.area {
  position: absolute;
  width: 100%;
  height: 100%;
  border-radius: 1rem;
  top: 0px;
  right: 0px;
  bottom: 0px;
  left: 0px;
}

.message {
  position: absolute;
  bottom: -30px;
  color: var(--color-red) !important;
}
