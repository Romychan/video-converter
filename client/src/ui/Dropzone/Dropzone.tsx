import {
  ChangeEvent,
  DragEvent,
  MutableRefObject,
  ReactElement,
  ReactNode,
  cloneElement,
  useRef,
  useState,
} from 'react';
import clsx from 'clsx';

import styles from './Dropzone.module.scss';

interface IDropzoneProps {
  title: string;
  icon?: ReactNode;
  uploadFileHandler: (file: FileList) => void;
  maxFileSize?: number;
  acceptedTypes?: string[];
  children: ReactElement;
}

export const Dropzone = ({
  title,
  icon,
  uploadFileHandler,
  maxFileSize = 1024,
  acceptedTypes = [],
  children,
}: IDropzoneProps) => {
  const inputRef = useRef() as MutableRefObject<HTMLInputElement>;
  const [dragActive, setDragActive] = useState(false);
  const [error, setError] = useState('');

  const isAcceptExt = (fileName: string) => {
    return acceptedTypes.includes(fileName.split('.').pop() as string);
  };

  const bytesToMegaBytes = (bytes: number) => bytes / (1024 * 1024);

  const handleDrag = (event: DragEvent) => {
    event.preventDefault();
    event.stopPropagation();

    if (event.type === 'dragenter' || event.type === 'dragover') {
      setDragActive(true);
    } else if (event.type === 'dragleave') {
      setDragActive(false);
    }
  };

  const handleDrop = (event: DragEvent) => {
    event.preventDefault();
    event.stopPropagation();
    setDragActive(false);

    if (event.dataTransfer.files && event.dataTransfer.files[0]) {
      handleNewFileUpload(event.dataTransfer.files);
    }
  };

  const handleUpload = () => {
    inputRef.current.click();
  };

  const handleNewFileUpload = (file: FileList) => {
    if (!isAcceptExt(file[0].name)) {
      setError('Ошибка загрузки. Данный формат файла не поддерживается');
      throw new Error('Ошибка валидации');
    }

    if (bytesToMegaBytes(file[0].size) > maxFileSize) {
      setError(
        `Ошибка загрузки. Размер файла не должен превышать ${maxFileSize} МБ`,
      );
      throw new Error('Ошибка валидации');
    }

    if (file.length) {
      uploadFileHandler(file);
    }
  };

  return (
    <div
      className={clsx(styles.container, {
        [styles.active]: dragActive,
      })}
    >
      <div
        onDragEnter={handleDrag}
        className={clsx(styles.dropzone, {
          [styles.error]: !!error,
        })}
        data-testid="dropzone-container"
      >
        <input
          type="file"
          className={styles.input}
          ref={inputRef}
          accept={'.' + acceptedTypes.join(', .')}
          onChange={(event: ChangeEvent<HTMLInputElement>) =>
            handleNewFileUpload(event.target.files as FileList)
          }
          data-testid="dropzone-input"
        />

        {icon ? <div className={styles.icon}>{icon}</div> : ''}

        <p className={clsx('subtitle', styles.subtitle)}>{title}</p>
        <p className={clsx('text text_small', styles.text)}>
          или нажмите на кнопку
        </p>

        {cloneElement(children, {
          onClick: handleUpload,
          className: clsx(styles.button),
        })}

        {dragActive && (
          <div
            className={styles.area}
            onDragEnter={handleDrag}
            onDragLeave={handleDrag}
            onDragOver={handleDrag}
            onDrop={handleDrop}
            data-testid="dropzone-zone"
          ></div>
        )}

        {!!error ? (
          <p
            className={clsx(styles.message, 'text text_small')}
            data-testid="dropzone-error"
          >
            {error}
          </p>
        ) : (
          ''
        )}
      </div>
      <div className={styles.background}></div>
    </div>
  );
};
