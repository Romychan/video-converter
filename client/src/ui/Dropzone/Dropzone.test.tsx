import {
  fireEvent,
  screen,
  render,
  waitFor,
  createEvent,
  RenderResult,
} from '@testing-library/react';
import { vi } from 'vitest';

import { Dropzone } from './Dropzone';

describe('Dropzone', () => {
  let container: RenderResult;
  const handleChange = vi.fn();

  beforeEach(() => {
    container = render(
      <Dropzone
        title="Test"
        uploadFileHandler={handleChange}
        acceptedTypes={['png']}
      >
        <button>Upload</button>
      </Dropzone>,
    );
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  it('should render correctly', () => {
    expect(screen.queryByText(/Test/i)).toBeInTheDocument();
    expect(screen.queryByText(/Upload/i)).toBeInTheDocument();
  });

  it('should upload file', async () => {
    const file = new File(['Filename'], 'Filename.png', { type: 'image/png' });
    const dropzoneInput = screen.getByTestId(
      'dropzone-input',
    ) as HTMLInputElement;

    await waitFor(() =>
      fireEvent.change(dropzoneInput, {
        target: { files: [file] },
      }),
    );

    expect(handleChange).toHaveBeenCalledTimes(1);
    expect(dropzoneInput.files?.[0]).toStrictEqual(file);
    expect(dropzoneInput.files).toHaveLength(1);
  });

  it('should drag and drop upload file', async () => {
    const file = new File(['Filename'], 'Filename.png', { type: 'image/png' });
    const dropzoneContainer = screen.getByTestId('dropzone-container');

    fireEvent.dragEnter(dropzoneContainer);

    const dropzoneZone = screen.getByTestId('dropzone-zone');
    const dropzoneZoneEvent = createEvent.drop(dropzoneZone);

    Object.defineProperty(dropzoneZoneEvent, 'dataTransfer', {
      value: {
        files: [file],
      },
    });
    fireEvent(dropzoneZone, dropzoneZoneEvent);

    expect(handleChange).toHaveBeenCalledTimes(1);
  });

  it('should handle error for not supported file', async () => {
    const file = new File(['Filename'], 'Filename.jpg', { type: 'image/jpg' });
    const dropzoneInput = screen.getByTestId(
      'dropzone-input',
    ) as HTMLInputElement;

    await waitFor(() =>
      fireEvent.change(dropzoneInput, {
        target: { files: [file] },
      }),
    );

    const dropzoneError = screen.getByTestId('dropzone-error');

    expect(dropzoneError).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    expect(container).toMatchSnapshot();
  });
});
