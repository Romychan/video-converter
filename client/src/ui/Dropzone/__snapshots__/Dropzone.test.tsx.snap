// Vitest Snapshot v1

exports[`Dropzone > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_container_c96111"
      >
        <div
          class="_dropzone_c96111"
          data-testid="dropzone-container"
        >
          <input
            accept=".png"
            class="_input_c96111"
            data-testid="dropzone-input"
            type="file"
          />
          <p
            class="subtitle _subtitle_c96111"
          >
            Test
          </p>
          <p
            class="text text_small _text_c96111"
          >
            или нажмите на кнопку
          </p>
          <button
            class="_button_c96111"
          >
            Upload
          </button>
        </div>
        <div
          class="_background_c96111"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_container_c96111"
    >
      <div
        class="_dropzone_c96111"
        data-testid="dropzone-container"
      >
        <input
          accept=".png"
          class="_input_c96111"
          data-testid="dropzone-input"
          type="file"
        />
        <p
          class="subtitle _subtitle_c96111"
        >
          Test
        </p>
        <p
          class="text text_small _text_c96111"
        >
          или нажмите на кнопку
        </p>
        <button
          class="_button_c96111"
        >
          Upload
        </button>
      </div>
      <div
        class="_background_c96111"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
